### ISSUE TITLE
HLSL - GetSamplePosition not implemented

### ISSUE THREAD

**Issue by gwihlidal** [KhronosGroup/glslang#753]
GetSamplePosition is not implemented on Texture2DMS resources.

i.e.
```
Texture2DMS<float> depthSrc;
uint sampleIdx = 1;
float val = depthSrc.GetSamplePosition(sampleIdx);
```

generates "error: '' : unimplemented: GetSamplePosition"

**Comment by steve-lunarg** [comment]
I recall looking into this a while back and not finding a SPIR-V equivalent.  (Which isn't to say there isn't one, only that I didn't find it).  The SamplePosition builtin did not appear sufficient: it's the position of the current sample, rather than an arbitrary sample number.

If there's a mapping, I can add this easily.

**Comment by steve-lunarg** [comment]
Related: #410.

**Comment by fluxlalonde** [comment]
There is no way in the current Vulkan/SPIR-V spec to get the non-standard sample positions.  If the pipeline is set up for standard positions, then we can generate code returning the position for the cases of rasterizationSamples <= 16. For the 32 and 64 cases there are no standard positions.

So we're left in the position of only being able to handle the 1, 2, 4, 8, and 16 sample cases for standard positions in the compiler. Non-standard layouts are un-knowable currently.

I would propose the addition of a SPIR-V instruction "OpImageQuerySamplePositions" taking a single integer parameter in the range of 0 to rasterizationSamples and return the position of that sample.

The driver has a relatively easy job, and could implement this by binding a static texture with all this information, and implement OpImageQuerySamplePositions as a simple 2D lookup on rasterizationSamles and the provided sample index (or 3D if they have multiple patterns available at a given sampling level, corresponding to a sampling pattern larger than a pixel).

In the short term, it seems easy enough (though a bit expensive) to generate SPIR-V code that returns the appropriate sample for the 1, 2, 4, 8, and 16 sample standard position cases, which will cover a significant amount of usage.

[gwihlidal's snippet above is a bit broken - its last line should be
`float2 pos = depthSrc.GetSamplePosition(sampleIdx);`
]

**Comment by steve-lunarg** [comment]
There's some code in #859 to implement the short term suggestion in @fluxlalonde's post just above.

As @fluxlalonde notes, it's ungainly.

**Comment by gwihlidal** [comment]
I think this should work in the interim (will report back during testing if that's not the case). And yes, thanks for fixing my transcription error, @fluxlalonde  :)

**Comment by arcady-lunarg** [comment]
The interim support has been implemented in #859 and it sounds like the proper solution will require going through the SPIR-V extension process first.
